### ISSUE TITLE
Cannot generate a complete CFG due to the function pointer 

### ISSUE THREAD

**Issue by daumbrella** [angr/angr#1005]
when I analyze a binary file with angr, I find that I cannot get a complete CFG due to the function pointer.
The following codes are an example

.text.tcpip_init:00000278 **tcpip_init**
.text.tcpip_init:00000278
.text.tcpip_init:00000278 var_18          = -0x18
.text.tcpip_init:00000278
.text.tcpip_init:00000278                 PUSH    {R0-R2,R4,R5,LR}
.text.tcpip_init:0000027A                 MOV     R5, R0
.text.tcpip_init:0000027C                 MOV     R4, R1
.text.tcpip_init:0000027E                 BL      lwip_init
.text.tcpip_init:00000282                 LDR     R3, =tcpip_init_done
.text.tcpip_init:00000284                 LDR     R0, =mbox
.text.tcpip_init:00000286                 STR     R5, [R3]
.text.tcpip_init:00000288                 LDR     R3, =tcpip_init_done_arg
.text.tcpip_init:0000028A                 MOVS    R1, #0
.text.tcpip_init:0000028C                 STR     R4, [R3]
.text.tcpip_init:0000028E                 BL      sys_mbox_new
.text.tcpip_init:00000292                 MOVS    R3, #2
.text.tcpip_init:00000294                 STR     R3, [SP,#0x18+var_18]
.text.tcpip_init:00000296                 MOVS    R2, #0
.text.tcpip_init:00000298                 MOV.W   R3, #0x300
.text.tcpip_init:0000029C                 LDR     R1, =(**tcpip_thread**+1)
.text.tcpip_init:0000029E                 LDR     R0, =aTcpIp     ; "tcp/ip"
.text.tcpip_init:000002A0                 BL      sys_thread_new
.text.tcpip_init:000002A4                 ADD     SP, SP, #0xC
.text.tcpip_init:000002A6                 POP     {R4,R5,PC}
.text.tcpip_init:000002A6 ; ---------------------------------------------------------------------------
.text.tcpip_init:000002A8 off_2A8         DCD tcpip_init_done     ; DATA XREF: tcpip_init+Ar
.text.tcpip_init:000002AC off_2AC         DCD mbox                ; DATA XREF: tcpip_init+Cr
.text.tcpip_init:000002B0 off_2B0         DCD tcpip_init_done_arg ; DATA XREF: tcpip_init+10r
.text.tcpip_init:000002B4 off_2B4         DCD **tcpip_thread+1**      ; DATA XREF: tcpip_init+24r
.text.tcpip_init:000002B8 off_2B8         DCD aTcpIp              ; DATA XREF: tcpip_init+26r
.text.tcpip_init:000002B8 ; .text.tcpip_init ends                 ; "tcp/ip"

If I generate a CFG from "tcpip_init" function, the CFG does not include the "tcpip_thread".
This program passes the "tcpip_thread"'s address as a parameter to the "sys_thread_new" function. In fact, "tcpip_thread" is called , but angr is not recognized. I know there is no obvious call statement of the "tcpip_thrad" in the function. Is this an indirect jump?
In short, what should I do if I want to add "tcpip_thread" to the control flow graphs?
[binaryfile.zip](https://github.com/angr/angr/files/2030149/binaryfile.zip)

**Comment by ltfish** [comment]
I think you are using CFGAccurate. It's the expected behavior under default settings of CFGAccurate. You can set `enable_function_hints=True` to account for function pointers in code when generating CFGAccurate.

This is not a problem for CFGFast (which is the default CFG). Here is my Python script:

```
import angr

p = angr.Project("tcpip.o", auto_load_libs=False)
cfg = p.analyses.CFG(show_progressbar=True)
print cfg.kb.functions["tcpip_thread"]
```

Execution output:

```
Function tcpip_thread [0x400001]
  Syscall: False
  SP difference: 0
  Has return: False
  Returning: True
  Arguments: reg: [], stack: []
  Blocks: [0x400001, 0x400017, 0x40001d, 0x400023, 0x40000f]
  Calling convention: None
```

`CFGFast` fails to resolve the jump table in `tcpip_thread`, which will be fixed soon.

**Comment by daumbrella** [comment]
**Thank you very much for your answer!**    
However, the execution output is still not the result I hoped. Although there is "tcpip_thread" function in the control flow graph generated by CFGFast, **there is still no call relationship between "tcpip_init" and "tcpip_thread".** The cfg generated by CFGFast is following. 
![tcpip_initcfg](https://user-images.githubusercontent.com/30562962/40415668-081c3a14-5eae-11e8-825d-3140ad38f8ec.PNG)

Maybe the jump table will be helpful. It is very worth looking forward to.

**Comment by ltfish** [comment]
There is no direct control flow transition from `tcp_init` to `tcp_thread`. In other words, `tcp_init` does not call `tcp_thread`. I'm not sure what you expect to see in the CFG.

**Comment by daumbrella** [comment]
Yes, you are right. There is no direct control flow from tcpip_init to tcpip_thread. While, in fact, after tcpip_thread's address is passed into the sys_thread_new, tcpip_thread is called.  And I also encountered a lot of similar situations that after function address pointer is passed, the function is indeed called. But because it is not a direct call, it cannot reveal the call relationship. 
I think it is possible to regard all function pointers as calls to this function? That means that every time I encounter a function pointer,  I add it to the successors of the calling function.
For this code, I just to generate a function call graph like this. 
![2](https://user-images.githubusercontent.com/30562962/40420295-8f335666-5eba-11e8-8a95-2d79fe577a95.PNG)

**Comment by ltfish** [comment]
You might want to add a SimProcedure for `sys_thread_new` and implement the `static_exits()` method for it. You may take a look at [__libc_start_main.py](https://github.com/angr/angr/blob/master/angr/procedures/glibc/__libc_start_main.py) for reference.
